### PR TITLE
Support restoring split backup files

### PR DESCRIPTION
**Pull request by jjermann**

### CODE REVIEW

**Comment by trailmax** [comment]
@jjermann Thank you for this PR, I was hoping to review it really quick but it looks more complex than I first thought. I'll give it a proper look tonight.

**Comment by jjermann** [comment]
> @jjermann Thank you for this PR, I was hoping to review it really quick but it looks more complex than I first thought. I'll give it a proper look tonight.

In retrospective it might have been better to let RestoreBackup do only one "restore" operation. I.e. instead of restoring a backup set + differential backups in one go (as in SSMS). It could be done with two RestoreBackup calls. That would "shift" the responsibility of "how to do that" to the caller though and would require some "finer" control on what operations are done exactly in each step...

If you prefer I could change it that way.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs [hunk 1]
```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
...
             {
-                var oldDbName = GetDatabaseName(backupFile, connection);
-                var newDatabaseName = settings.NewDatabaseName ?? oldDbName;
-                context.Log.Information($"Using database name '{newDatabaseName}' to be a name for the restored database");
+                bool hasDifferentialBackup = differentialBackupFiles != null && differentialBackupFiles.Any();
+                var mainCommand = GetSqlCommand(context, connection, settings, backupFiles, settings.BackupSetFile, hasDifferentialBackup);
+                mainCommand.ExecuteNonQuery();
+                if (hasDifferentialBackup)
+                {
+                    var differentialCommand = GetSqlCommand(context, connection, settings, differentialBackupFiles, settings.DifferentialBackupSetFile, false, false);
```
Comment (trailmax) [inline src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: It would help the readability to name the optional parameters.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs [hunk 2]
```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
...
-        internal static void RestoreSqlBackup(ICakeContext context, String connectionString, FilePath backupFile, RestoreSqlBackupSettings settings)
+        internal static void RestoreSqlBackup(ICakeContext context, String connectionString, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, IList<FilePath> differentialBackupFiles = null)
         {
             using (var connection = SqlServerAliasesImpl.OpenSqlConnection(context, connectionString))
             {
-                var oldDbName = GetDatabaseName(backupFile, connection);
-                var newDatabaseName = settings.NewDatabaseName ?? oldDbName;
-                context.Log.Information($"Using database name '{newDatabaseName}' to be a name for the restored database");
+                bool hasDifferentialBackup = differentialBackupFiles != null && differentialBackupFiles.Any();
+                var mainCommand = GetSqlCommand(context, connection, settings, backupFiles, settings.BackupSetFile, hasDifferentialBackup);
```
Comment (trailmax) [inline src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: It would help the readability to name the optional parameters.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs [hunk 3]
```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
...
-                    if (i < logicalNames.Count - 1)
-                    {
-                        sql += ", \r\n"; // only need comma before penultimate list
-                    }
+                    sql += ",\r\n"; // only need comma before penultimate list
                 }
-                if (settings.WithReplace)
+            }
+
+            if (settings.WithReplace)
```
Comment (trailmax) [inline src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: Replace is not really applicable to diff backups, yet it will crop in if specified in options

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs [hunk 4]
```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
...
+                if (hasDifferentialBackup)
+                {
+                    var differentialCommand = GetSqlCommand(context, connection, settings, differentialBackupFiles, settings.DifferentialBackupSetFile, false, false);
+                    differentialCommand.ExecuteNonQuery();
+                }
+            }
+        }
 
-                var logicalNames = GetLogicalNames(backupFile, connection);
+        private static SqlCommand GetSqlCommand(ICakeContext context, SqlConnection connection, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, int backupSetFile = 1, bool noRecovery = false, bool allowAlterDb = true)
```
Comment (trailmax) [inline src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: Given amount of parameters, optional parameters and if-statements, I strongly feel that this method should be split:
`CommandForSingleUserMode`
`CommandForFullBackups`
`CommandForDiffBackups`
`CommandForRestoringMultiUserMode`

Otherwise we are doing too much in one place and that is hard to follow and prone to bugs. Also full/diff backups do not always have the same options available, so these are really different SQL statements.

I'm sure we'll be able to extract methods to populate file names and logical names parameter values so these can be reused for Full/Diff backups. 

After that is done, I'm sure all the optional parameters will be gone and amount of if-statements will be reduced.

File: src/Tests/Tests.csproj [hunk 5]
```diff
@@ -109,9 +109,27 @@
     <None Include="app.config">
       <SubType>Designer</SubType>
     </None>
+    <None Include="differentialMultiFilesBackup1.bak">
```
Comment (trailmax) [inline src/Tests/Tests.csproj]: Can you possibly move all the bak files to a folder? to hide them away in Solutions Explorer

File: ReleaseNotes.md [hunk 6]
```diff
@@ -1,3 +1,6 @@
+### New in 2.0.2
```
Comment (trailmax) [inline ReleaseNotes.md]: I'd say that's a pretty significant new feature, so it deserves 2.1.0 version

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 4 above)
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: Ok, I'll do an update. Are you ok to break compatibility to older versions e.g. regarding method signatures, behaviors and/or RestoreSqlBackupSettings?

Or was the above suggestion mostly meant for the "internal" design?

File: ReleaseNotes.md (hunk 6 above)
Comment (jjermann) [reply ReleaseNotes.md]: How does that exactly work? In the code I saw that git version is used to generate the AssemblyInfo content.

Comment (trailmax) [reply ReleaseNotes.md]: I'll put a tag "2.1.0" on a git commit that will be a version realesed. And this file is edited by hand - so the numbers will have to coordinate

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 4 above)
Comment (trailmax) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: I'd rather not break the compatibility. What I meant is for the internal implemenatation.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs [hunk 7]
```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
...
+            }
+
+            if (noRecovery)
+            {
+                sql += ",\r\n norecovery;";
+            }
+            else
+            {
+                sql += $",\r\n recovery;";
+                if (allowAlterDb)
```
[the remark this answers is not in the document]
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: After some thought I believe it's ok to have several parameters for GetRestoreSqlBackupCommand because a single restore database command can (in general) have several options. However it should definitely have only that responsibility (in particular it shouldn't manage argument parsing or control logic).

All argument decisions and control logic is now done in RestoreSqlBackup.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 1 above)
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: See my remark above -> I changed the internal design.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 2 above)
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: See my remark above -> I changed the internal design.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 3 above)
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: The specifics of the parameters of the individual database restore commands are now no longer handled inside `GetRestoreSqlBackupCommand`. This also solves the issue mentioned here.

File: src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs (hunk 4 above)
Comment (jjermann) [reply src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs]: > 
> 
> I'd rather not break the compatibility. What I meant is for the internal implemenatation.

All right. I tried to improve the design a bit.

File: src/Tests/Tests.csproj (hunk 5 above)
Comment (jjermann) [reply src/Tests/Tests.csproj]: Done.

**Comment by jjermann** [comment]
Unfortunately there seem to be some cake unit test issues now (?).

File: src/Cake.SqlServer/Backup/BackupAliases.cs [hunk 8]
```diff
@@ -101,5 +144,33 @@ public static void BackupDatabase(this ICakeContext context, string connectionSt
...
+        ///             var connString = @"data source=(localdb)\MSSqlLocalDb";
+        ///             var databaseName = "MyDatabase";
+        ///             SetDatabaseSingleUserMode(connString, databaseName, true);
+        ///             // SomeTask requiring single user mode
+        ///             SetDatabaseSingleUserMode(connString, databaseName, false);
+        ///         });
+        /// </code>
+        /// </example>
+        [CakeMethodAlias]
+        public static void SetDatabaseSingleUserMode(this ICakeContext context, String connectionString, String databaseName, bool singleUserMode)
```
[the remark this answers is not in the document]
Comment (jjermann) [reply src/Cake.SqlServer/Backup/BackupAliases.cs]: Done.

**Comment by trailmax** [comment]
MERGED!
Thank you for your work!

**Comment by trailmax** [comment]
I'm hoping I'll get some time this week to get a release out

**Comment by trailmax** [comment]
@jjermann I'm running into version mismatch between LocalDB version locally and on AppVeyor. I'm getting exception `The database was backed up on a server running version 13.00.5081. That version is incompatible with this server, which is running version 12.00.4100. Either restore the database on a server that supports the backup, or use a backup that is compatible with this server.
RESTORE DATABASE is terminating abnormally.`

And I can't seem to be able to use any later version of LocalDB on the build server.
Would you be able to provide backup files that were created by LocalDB v12?

**Comment by trailmax** [comment]
@jjermann see this build: https://ci.appveyor.com/project/trailmax/cake-sqlserver/builds/28195472/tests

**Comment by jjermann** [comment]
I don't know exactly how to create backups for LocalDB v12. I used a +- recent MSSQL.
But all I did was:
1. I restored the original .bak.
2. I created a backup but for 3 files instead of just 1 (multiFilesBackup*.bak).
3. I did one simple (not necessary but potentially testable) change.
4. I created a differential backup for 3 files (differentialMultiFilesBackup*.bak).

If you do this on LocalDB v12 and then replace the .bak files by the new ones it should still all work.

Alternatively: Do you know what I need to do in MSSQL to produce valid (multifile + differential) backups?

**Comment by trailmax** [comment]
@jjermann do you have scripts to create multi-file backups? I can run it here.

**Comment by jjermann** [comment]
> @jjermann do you have scripts to create multi-file backups? I can run it here.

I can check next week. I believe something like:
`
BACKUP DATABASE [foo] TO 
DISK = 'C:\bar1.bak', 
DISK = 'C:\bar2.bak', 
DISK = 'C:\bar3.bak' 
WITH INIT , NOUNLOAD , NAME = 'foo2', NOSKIP , STATS = 10, NOFORMAT
`
and for differential use `WITH DIFFERENTIAL`.

**Comment by trailmax** [comment]
THat's the release is on Nuget.org. Currently validating, but should be live within a few hours.
